server/accounts.go

```diff
@@ -863,140 +1026,137 @@ func (a *Account) removeServiceImport(subject string) {
+// This checks for any response map entries. If you specify an si we will only match and
+// clean up for that one, otherwise we remove them all.
+func (a *Account) checkForReverseEntry(reply string, si *serviceImport, checkInterest bool) {
 	a.mu.RLock()
-	defer a.mu.RUnlock()
-	return a.maxaettl
-}
+	if len(a.imports.rrMap) == 0 {
+		a.mu.RUnlock()
+		return
+	}
 
-// SetAutoExpireTTL sets the ttl for response maps.
-func (a *Account) SetAutoExpireTTL(ttl time.Duration) {
-	a.mu.Lock()
-	defer a.mu.Unlock()
-	a.maxaettl = ttl
-}
+	if subjectHasWildcard(reply) {
+		a.mu.RUnlock()
+		go a.checkForReverseEntries(reply, checkInterest)
+		return
+	}
 
-// Return a list of the current autoExpireResponseMaps.
-func (a *Account) autoExpireResponseMaps() []*serviceImport {
-	a.mu.RLock()
-	if len(a.imports.services) == 0 {
+	sres := a.imports.rrMap[reply]
+	if sres == nil {
 		a.mu.RUnlock()
-		return nil
+		return
 	}
-	aesis := make([]*serviceImport, 0, len(a.imports.services))
-	for _, si := range a.imports.services {
-		if si.ae {
-			aesis = append(aesis, si)
+
+	// If we are here we have an entry we should check.
+	// If requested we will first check if there is any
+	// interest for this subject for the entire account.
+	// If there is we can not delete any entries yet.
+	// Note that if we are here reply has to be a literal subject.
+	if checkInterest {
+		rr := a.sl.Match(reply)
+		// If interest still exists we can not clean these up yet.
+		if len(rr.psubs)+len(rr.qsubs) > 0 {
+			a.mu.RUnlock()
+			return
 		}
 	}
-	sort.Slice(aesis, func(i, j int) bool {
-		return aesis[i].ts < aesis[j].ts
-	})
-
 	a.mu.RUnlock()
-	return aesis
-}
-
-// MaxResponseMaps return the maximum number of
-// non auto-expire response maps we will allow.
-func (a *Account) MaxResponseMaps() int {
-	a.mu.RLock()
-	defer a.mu.RUnlock()
-	return int(a.maxnrm)
-}
 
-// SetMaxResponseMaps sets the max outstanding non auto-expire response maps.
-func (a *Account) SetMaxResponseMaps(max int) {
+	// Delete the appropriate entries here based on optional si.
 	a.mu.Lock()
-	defer a.mu.Unlock()
-	a.maxnrm = int32(max)
+	if si == nil {
+		delete(a.imports.rrMap, reply)
+	} else {
+		// Find the one we are looking for..
+		for i, sre := range sres {
+			if sre.msub == si.from {
+				sres = append(sres[:i], sres[i+1:]...)
+				break
+			}
+		}
+		if len(sres) > 0 {
+			a.imports.rrMap[si.to] = sres
+		} else {
+			delete(a.imports.rrMap, si.to)
+		}
+	}
+	a.mu.Unlock()
+
+	// If we are here we no longer have interest and we have
+	// response entries that we should clean up.
+	if si == nil {
+		for _, sre := range sres {
+			acc := sre.acc
+			var trackingCleanup bool
+			var rsi *serviceImport
+			acc.mu.Lock()
+			if rsi = acc.exports.responses[sre.msub]; rsi != nil {
+				delete(acc.exports.responses, rsi.from)
+				trackingCleanup = rsi.tracking && rsi.rc != nil
+			}
+			acc.mu.Unlock()
+
+			if trackingCleanup {
+				acc.sendReplyInterestLostTrackLatency(rsi)
```

Comment:
Something is up here, with this line active the only metric I get is a 408 even with requests that worked fine.  Comment it out I and I do get the correct latency sample for a full request.

Still trying to figure out how this works - but some logic issue around here.

Reply:
Will take a look.